Add --nondet-static-except flag to CBMC

mmuesly opened the pull request:
--nondet-static-except behaves as --nondet-static but allows to filter
the symbols which are effected by --nondet-static. Every symbol
that is effected by nondet-static and is not in the comma separated
symbol list passed to --nondet-static-except will be nondet when
using this flag. All symbols in the list keep their concret initalization.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: f3502ff69956b84bc8dbf09c20e46b4871080ddb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112499997
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

mmuesly commented:
@allredj could you give me an advise regarding what I should do to pass the compatibility check?
I can not access the linked build.

allredj commented:
@mmuesly Hi! You can disregard the compatibility check for now. Actually I've just disabled it because TG and CBMC are already incompatible at this time.

smowton commented:
Could I suggest for symmetry we should have `--nondet-static-variables` and `--nondet-static-exclude-variables`, with the rule "nondet if blanket nondet-static or nondet-static-variables AND not nondet-static-exclude-variables"?

peterschrammel commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -1006,7 +1021,8 @@ void cbmc_parse_optionst::help()
     // NOLINTNEXTLINE(whitespace/line_length)
-    " --nondet-static              add nondeterministic initialization of variables with static lifetime\n"
+    " --nondet-static              add nondeterministic initialization of variables with static lifetime\n" // NOLINT(*)
+    " --nondet-static-except e     same as nondet-static except variables listed in e\n" // NOLINT(*)
```
As @smowton, I have a preference for the suffix `-exclude`. Also, comma-separation should not be used, but multiple instances of the option should be supplied instead.

peterschrammel commented on regression/cbmc/nondet-static-except_subfolder/main.c:
```diff
@@ -0,0 +1,9 @@
+  method1();
+  method2();
+  return 0;
+}
```
missing newline

kroening commented:
This appears to be a feature that should go into the harness generator?

peterschrammel commented on regression/cbmc/nondet-static-except_subfolder/test.desc:
```diff
@@ -0,0 +1,12 @@
+CORE
+package2/file.c
+--nondet-static-except package1/file.c:myVal,package2/file.c:myVal2 main.c package1/file.c 
```
```suggestion
--nondet-static-exclude package1/file.c:myVal --nondet-static-exclude package2/file.c:myVal2 main.c package1/file.c 
```

karkhaz commented:
@kroening to clarify, are you asking for @mmuesly to do additional work to integrate this into the harness generator, or was your comment just a note to Diffblue people?

karkhaz commented:
@smowton not sure if I totally understand the semantics that you're proposing, but I see a potential for confusion here: what happens if a user passes both `--nondet-static` and `--nondet-static-variables-exclude`, but no `--nondet-static-variables`?

Your wording seems to imply that the exclude list would be entirely ignored, because `--nondet-static` makes everything nondet. On the other hand, if `--nondet-static` does in fact take into account the exclude list, then I don't see why it's any different to `--nondet-static-variables`.

The concrete reason why this is a concern is: we use a "common" Makefile that sets up most of the options to CBMC (like `--nondet-static`), using a variable like `CBMCFLAGS`. Then, for individual CBMC tests, we have a local makefile that appends to `CBMCFLAGS`---this is where we would set `--nondet-static-variables-except`. The local Makefile then runs `cbmc $(CBMCFLAGS) foo.c`. With your proposal, it seems very easy for a user to set `--nondet-static` (rather than `--nondet-static-variables`) at the global level, which would then silently override the exclude list.

If you really prefer the symmetry, I propose that we make `--nondet-static-variables` an alias for `--nondet-static`. Maybe even remove `--nondet-static` from all documentation, but keep it in the codebase so that we don't break existing users of that flag.

mmuesly commented:
I changed the PR as requested by @peterschrammel.
Further, I made sure that `--nondet-static` does not overwrite `--nondet-static-exclude`. I missed this side effect in the first draft. As described by @karkhaz, this is important for our use case.

kroening commented:
@karkhaz The PR should change to make this feature an addition to the harness generator (unless I am misunderstanding the purpose).
All features that produce a harness for a given piece of code should go there.

karkhaz commented:
@kroening this feature doesn't generate new harnesses. The new flag gives users finer control over what global variables are initialized to nondet. This flag can be used on arbitrary goto-binaries, so it's not related to harness generation.

mmuesly commented:
@kroening This PR is intended to modify the behavior of an existing CBMC flag and is a flag CBMC borrows from goto-instrument. It is adding only a couple of lines to the existing function implementation. Do you plan to move the original code from goto-instrument to the goto-harness package?

kroening commented:
@karkhaz if it's about transforming an existing program, it should go into goto-instrument!
cbmc should not get any options that change semantics of the goto program.

kroening commented:
@mmuesly The --nondet-static flag is indeed an issue; ideally, it should be removed from cbmc, which is at least hinted by the fact that its implementation lives in goto-instrument. The reason why it's still there is that there are existing users.

karkhaz commented:
@kroening OK. So I suggest that @mmuesly does the following:
1. Add the new flag to goto-instrument and remove it from cbmc
1. Move the tests to the goto-instrument regression directory. You won't
   have to change the tests very much, because (looking at
   `regressin/goto-instrument/chain.sh`) the tests in that directory have
   both `goto-instrument` and `cbmc` run over them. So you would use
   `goto-instrument` to instrument the binaries with
   `--nondet-static-exclude`, and then use CBMC to check that the
   verification result is as expected.

As a *separate* PR, we might consider removing `--nondet-static` from
the help output of CBMC (but keeping it in the actual codebase, to
prevent breaking legacy users).

mmuesly committed:
Add --nondet-static-except flag to goto-instrument

--nondet-static-except behaves as --nondet-static but allows to filter
the symbols which are effected by --nondet-static. Every symbol
that is effected by nondet-static and is not in the comma separated
symbol list passed to --nondet-static-except will be nondet when
using this flag. All symbols in the list keep their concret initalization.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: dbda7c01c9f767974c21d337cbb863b5628759c6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114465285

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 585fdff016c43eef6e9d8a3ea8f94c528f098509).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114480311
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5cd8b252b3cc68c3274363814135fd42cad58940).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/114523926

mmuesly commented:
@peterschrammel I implemented @karkhaz suggestions.
What should I do next?
Should I open a second PR for removing nondet-static from the cbmc help output?
That is from my point of view the last open point no yet addressed raised in this discussion.

tautschnig commented on regression/goto-instrument/chain.sh:
```diff
@@ -8,13 +8,23 @@ cbmc=$3
 name=${*:$#}
-name=${name%.c}
+
+if [[ ! ($name =~ " ") ]];  then
```
Would it be easier to use `if [[ x$name != x ]] ; then` ?

tautschnig commented on regression/goto-instrument/chain.sh:
```diff
@@ -8,13 +8,23 @@ cbmc=$3
 args=${*:5:$#-5}
 
-if [[ "${is_windows}" == "true" ]]; then
+if [[ "${is_windows}" == "true" && ("${name}" =~ " ") ]]; then
```
As above - matching on whitespace is just a bit odd.

tautschnig commented on regression/goto-instrument/chain.sh:
```diff
@@ -8,13 +8,23 @@ cbmc=$3
+elif [[ "${is_windows}" == "true" ]]; then
   $goto_cc "${name}.c"
   mv "${name}.exe" "${name}.gb"
+elif [[ ("${name}" =~ " ") ]]; then
```
As above

tautschnig commented on regression/goto-instrument/nondet-static-exclude_subfolder/package1/file.c:
```diff
@@ -0,0 +1,10 @@
+#include "assert.h"
+#include "stdint.h"
```
`#include <assert.h>`, `#include <stdint.h>` unless you really want to have the option to use a local copy of those files?

tautschnig commented on regression/goto-instrument/nondet-static-exclude_subfolder/package1/file.c:
```diff
@@ -0,0 +1,10 @@
+#include "assert.h"
+#include "stdint.h"
+static uint32_t myVal = 3;
+static uint32_t myVal2 = 4;
```
Is the use of `uint32_t` relevant for this test? Wouldn't `unsigned` or even `int` do?

tautschnig commented on regression/goto-instrument/nondet-static-exclude_subfolder/package2/file.c:
```diff
@@ -0,0 +1,10 @@
+#include "assert.h"
+#include "stdint.h"
+static uint32_t myVal = 3;
+static uint32_t myVal2 = 4;
```
As above

tautschnig commented on regression/goto-instrument/nondet_static_exclude/test.desc:
```diff
@@ -0,0 +1,18 @@
+^VERIFICATION FAILED$
+^EXIT=10$
+^SIGNAL=0$
+assertion test1 == (?:NULL||0): SUCCESS
```
Why such elaborate pattern matching?

mmuesly replied:
Because in some test in CI this was NULL and in other it was 0.

What would you expect to be the default value in CI for all compilers?

tautschnig replied:
That seems fine, but wouldn't then `(NULL|0)` do instead of `(?:NULL||0)`?

mmuesly replied:
Yes

tautschnig commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1607,6 +1616,7 @@ void goto_instrument_parse_optionst::help()
     " --isr <function>             instruments an interrupt service routine\n"
     " --mmio                       instruments memory-mapped I/O\n"
     " --nondet-static              add nondeterministic initialization of variables with static lifetime\n" // NOLINT(*)
+    " --nondet-static-exclude e    same as nondet-static except for the variable e (use multiple times if required)\n" // NOLINT(*)
```
This is a very long line (also in the output!), please line-wrap

tautschnig commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -136,3 +140,68 @@ void nondet_static(goto_modelt &goto_model)
+  for(optionst::value_listt::const_iterator i = except_values.begin();
+      i != except_values.end();
+      ++i)
+  {
```
I'd suggest to use a ranged for: `for(const auto &except : except_values)`

tautschnig commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -136,3 +140,68 @@ void nondet_static(goto_modelt &goto_model)
+    bool prefix_found = false;
+    for(size_t j = 0; j < i->length(); ++j)
+    {
+      if((*i)[j] == ':')
```
Use `i->find(':')`?!

tautschnig commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -136,3 +140,68 @@ void nondet_static(goto_modelt &goto_model)
+    }
+  }
+
+  symbol_tablet symbol_table = ns.get_symbol_table();
```
Why is a copy taken?

mmuesly replied:
Because namespacet has a const symbol_table.
If I take a const reference the iterators are also const and I cannot call get_writable_symbol on the iterator. Which makes sense from my point of view that a const symbol_table must not be modified. Anyhow, I want to modify it. Therefore, I copy it and then modify the copy.

Eventually, I have overseen a possibility to this without the copy. But I don’t know how to get a symbol_tablet::iteratort instead of the symbolst::const_iterator if I have a const reference to a symbol table.

Any advice?

tautschnig replied:
You don't really need to go through `ns`, because the namespace itself you only constructed earlier from `goto_model.symbol_table` - I think you could just use that symbol table directly?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b1450414a60e144b319d4cd92e987a40fb6e32ce).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115658522

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 362ab4de235c4560e5567717827cf1eb5eef1e4d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115660321

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: afd0b0abbc6f272b27f78a00b76beeaef6aab97d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/115662790

mmuesly commented:
@tautschnig Thank you for the review and comments. I fixed all build issues by now and addressed all your comments.

romainbrenguier commented on src/goto-instrument/nondet_static.h:
```diff
@@ -20,6 +20,8 @@ Date: November 2011
 #ifndef CPROVER_GOTO_INSTRUMENT_NONDET_STATIC_H
 #define CPROVER_GOTO_INSTRUMENT_NONDET_STATIC_H
 
+#include <util/options.h>
```
you can use a forward declaration instead if you use `std::list<std::string>` instead of `optionst::value_listt` below

mmuesly replied:
Unless there is a plan to drop the `optionst::value_listt` in the long term, I personally think using explicitly this type makes it clearer what is expected as a parameter.

romainbrenguier replied:
I personally disagree that it is clearer (the name value_listt carries strictly less information than list<string>), but I know people can have a different opinion on that so I won't insist.

romainbrenguier commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -136,3 +140,64 @@ void nondet_static(goto_modelt &goto_model)
+        to_exclude.insert("./" + except);
+      }
+    }
+    if(!prefix_found)
```
replace by `else`

romainbrenguier commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -136,3 +140,64 @@ void nondet_static(goto_modelt &goto_model)
+  for(auto const &except : except_values)
+  {
+    bool prefix_found = false;
+    auto it = except.find(":");
```
`const bool prefix_found = except.find(":") != std::string::npos;`

romainbrenguier commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -136,3 +140,64 @@ void nondet_static(goto_modelt &goto_model)
+    {
+      prefix_found = true;
+      to_exclude.insert(except);
+      if(except.length() > 2 && except[0] == '.' && except[1] == '/')
```
use `has_prefix`

romainbrenguier commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -136,3 +140,64 @@ void nondet_static(goto_modelt &goto_model)
+      }
+      else
+      {
+        to_exclude.insert("./" + except);
```
I don't really understand this piece of code, if `except` is "./path" we insert "path" and if it is "path" we insert "./path"?

mmuesly replied:
Yes. The location path returned by the namespace symbol after lookup sometimes starts with `./` and sometimes not. To make the lookup more deterministic, I add the prefix eventually to the passed parameter or remove it. So the exclude list always contains `path` and `./path` as an excluded location.

romainbrenguier replied:
Ok I see. Maybe add a comment about that. Ideally I would expect the path to be stored in some canonical form (but I guess it's alright for now)

romainbrenguier replied:
Ok I see. Maybe add a comment about that. Ideally I would expect the path to be stored in some canonical form (but I guess it's alright for now)

romainbrenguier commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -136,3 +140,64 @@ void nondet_static(goto_modelt &goto_model)
+  }
+
+  // Update the symbol_table in the model
+  goto_model.symbol_table = symbol_table;
```
not needed, symbol table is already a reference to `goto_model.symbol_table`

mmuesly commented:
@peterschrammel and @romainbrenguier, thank you for the reviews. I addressed romainbrenguier's comments.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8e158b5a57af04100f84a4eb3e1b80b1020fd6a6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/116671062

romainbrenguier commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -136,3 +140,59 @@ void nondet_static(goto_modelt &goto_model)
+
+  for(symbol_tablet::iteratort symbol_it = symbol_table.begin();
+      symbol_it != symbol_table.end();
+      symbol_it++)
```
Range for if possible (otherwise ++symbol_it and you can use auto for the iterator type)

romainbrenguier commented on src/goto-instrument/nondet_static.cpp:
```diff
@@ -136,3 +140,59 @@ void nondet_static(goto_modelt &goto_model)
+      symbol_it++)
+  {
+    symbolt &symbol = symbol_it.get_writeable_symbol();
+    std::string qualified_name = id2string(symbol.location.get_file()) + ":" +
```
Const
